### PR TITLE
PDA Ringtone Preference

### PR DESCRIPTION
**Pull request by GoldenAlpharex**
## About The Pull Request
This PR aims to make it easier for players to have their own custom PDA ringtones by making in a prefs menu preference, so that they don't have to manually set it every single shift before getting their very first PDA message.

To do that, I had to properly implement the concept of a text-based preference that wasn't just names, so from there emerges `/datum/preference/text`, made to be as simple to use as possible. Then, I added a new TGUI component for it, `FeatureShortTextInput` (short, to leave the door opened for a non-short text input component for things that might require the player to write a bit more, by making it multiline, for instance), which is also going to prevent the player from going above the maximum length defined by the text preference. Do note that the default max text length won't fit entirely in the box without scrolling to the right, but I honestly just didn't know what value to put there so I set it to something that sounded reasonable, so 256.

As of right now, that preference is only applied to humans, and not to silicons, mostly because it would possibly be better off as a different preference for those, much like how Cyborg, AI and real names are three different preferences.

## Why It's Good For The Game
It reduces the tedium of having a custom ringtone by making it a preference, there's not really any added value to the game to have to manually enter the same value in the same input field every single round. It also should incentivize the use of a relatively neat feature that usually barely sees any use.

## Changelog

:cl: GoldenAlpharex
qol: Added a preference for PDA ringtones, to save a custom ringtone that will then be applied automatically to your character when spawned in the game. Gone are the days of having to do it manually at the start of every round!
/:cl:

### CODE REVIEW

File: code/modules/client/preferences/_preference.dm [hunk 1]
```diff
@@ -542,3 +542,24 @@ GLOBAL_LIST_INIT(preference_entries_by_key, init_preference_entries_by_key())
...
+/// A string-based preference accepting arbitrary string values entered by the user, with a maximum length.
+/datum/preference/text
+	abstract_type = /datum/preference/text
+
+	/// What is the maximum length of the value allowed in this field?
+	var/maximum_value_length = 256
+
+
+/datum/preference/text/deserialize(input, datum/preferences/preferences)
+	return STRIP_HTML_SIMPLE(input, maximum_value_length)
```
Comment (Mothblocks) [inline code/modules/client/preferences/_preference.dm:556]: Whether or not to strip HTML should be up to consumers, IMO, I can think of plenty of cases where I wouldn't want this. It's fine as a variable.

File: code/modules/client/preferences/_preference.dm [hunk 2]
```diff
@@ -542,3 +542,24 @@ GLOBAL_LIST_INIT(preference_entries_by_key, init_preference_entries_by_key())
...
+
+/datum/preference/text/create_default_value()
+	return ""
+
+/datum/preference/text/is_valid(value)
+	return istext(value)
```
Comment (Mothblocks) [inline code/modules/client/preferences/_preference.dm:562]:
```suggestion
	return istext(value) && value.len < maximum_value_length
```

File: code/modules/client/preferences/pda_ringtone.dm [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/datum/preference/text/pda_ringtone
```
Comment (Mothblocks) [inline code/modules/client/preferences/pda_ringtone.dm:1]: Document it

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/base.tsx [hunk 4]
```diff
@@ -3,7 +3,7 @@ import { BooleanLike, classes } from 'common/react';
 import { ComponentType, createComponentVNode, InfernoNode } from 'inferno';
 import { VNodeFlags } from 'inferno-vnode-flags';
 import { sendAct, useBackend, useLocalState } from '../../../../backend';
-import { Box, Button, Dropdown, NumberInput, Stack } from '../../../../components';
+import { Box, Button, Dropdown, NumberInput, Stack, Input } from '../../../../components';
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/base.tsx:6]: ABC order, this should lint.

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/character_preferences/pda.tsx [hunk 5]
```diff
@@ -13,3 +13,10 @@ export const pda_style: Feature<string> = {
   description: 'The style of your equipped PDA. Changes font.',
   component: FeatureDropdownInput,
 };
+
+export const pda_ringtone: Feature<string> = {
+  name: 'PDA Ringtone',
+  description:
+    'Want your PDA to say something other than "beep"? Accepts the first 20 characters.',
```
Comment (Mothblocks) [inline tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/character_preferences/pda.tsx:20]: What does "Accepts the first 20 characters." mean? Is this something made redundant with proper max length filtering?

Comment (Mothblocks) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/character_preferences/pda.tsx:20]:
```suggestion
    "The ringtone you'll hear when someone sends you a PDA message.",
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/client/preferences/_preference.dm (hunk 1 above)
Comment (GoldenAlpharex) [reply code/modules/client/preferences/_preference.dm:556]: Sure yeah.

File: tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/character_preferences/pda.tsx (hunk 5 above)
Comment (GoldenAlpharex) [reply tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/character_preferences/pda.tsx:20]: True, I copy-pasted this from the old component we had downstream, which hadn't implemented the character limit properly like I did, I'll fix that yeah.

File: code/modules/jobs/job_types/_job.dm [hunk 6]
```diff
@@ -155,6 +155,16 @@
 	if(!ishuman(spawned))
 		return
 
+	var/mob/living/carbon/human/spawned_human = spawned
+
+	// Look for PDA in belt or left pocket to update its ringtone based on our client's prefs.
+	var/obj/item/modular_computer/tablet/pda/pda = spawned_human.belt
```
Comment (JohnFulpWillard) [inline code/modules/jobs/job_types/_job.dm:161]: ``/datum/outfit/job/post_equip`` already does some PDA stuff and does it in a pretty nice way that doesn't have hardcoded paths like you have here, do you think you can put it there instead?

Comment (GoldenAlpharex) [reply code/modules/jobs/job_types/_job.dm]: I figured out a way to do it,

File: code/modules/client/preferences/_preference.dm [hunk 7]
```diff
@@ -542,3 +542,27 @@ GLOBAL_LIST_INIT(preference_entries_by_key, init_preference_entries_by_key())
...
+
+	/// Should we strip HTML the input or simply restrict it to the maximum_value_length?
+	var/should_strip_html = TRUE
+
+
+/datum/preference/text/deserialize(input, datum/preferences/preferences)
+	return should_strip_html ? STRIP_HTML_SIMPLE(input, maximum_value_length) : copytext(input, 1, maximum_value_length)
+
+/datum/preference/text/create_default_value()
+	return ""
```
Comment (JohnFulpWillard) [inline code/modules/client/preferences/_preference.dm:562]:
```suggestion
	return MESSENGER_RINGTONE_DEFAULT
```

Comment (GoldenAlpharex) [reply code/modules/client/preferences/_preference.dm:562]: This isn't the PDA preferences, this is the barebones text preference, it absolutely shouldn't have that as a default value .-.

**Pull request merged**